File: agentlang/adapters/openai.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Callable
from urllib import error, request


class OpenAIAdapterError(RuntimeError):
    pass


@dataclass(frozen=True)
class OpenAIResponsesClient:
    api_key: str
    base_url: str = "https://api.openai.com/v1"
    timeout_s: float = 45.0

    def complete(
        self,
        *,
        model: str,
        prompt: str,
        system: str | None = None,
        max_output_tokens: int | None = None,
    ) -> str:
        data = self.create_response(
            model=model,
            input_items=self._build_input(prompt=prompt, system=system),
            max_output_tokens=max_output_tokens,
        )
        text = _extract_text(data)
        if not text:
            raise OpenAIAdapterError("OpenAI response contained no text output.")
        return text

    def complete_with_tools(
        self,
        *,
        model: str,
        prompt: str,
        system: str | None,
        tools: list[dict[str, Any]],
        call_tool: Callable[[str, dict[str, Any]], Any],
        max_output_tokens: int | None = None,
        max_round_trips: int = 8,
    ) -> str:
        response = self.create_response(
            model=model,
            input_items=self._build_input(prompt=prompt, system=system),
            tools=tools,
            parallel_tool_calls=False,
            max_output_tokens=max_output_tokens,
        )

        for _ in range(max_round_trips):
            function_calls = _extract_function_calls(response)
            if not function_calls:
                text = _extract_text(response)
                if not text:
                    raise OpenAIAdapterError("OpenAI response contained no text output.")
                return text

            output_items: list[dict[str, Any]] = []
            for tool_call in function_calls:
                name = tool_call["name"]
                call_id = tool_call["call_id"]
                arguments_text = tool_call["arguments"]
                try:
                    arguments = json.loads(arguments_text)
                except json.JSONDecodeError as exc:
                    raise OpenAIAdapterError(
                        f"OpenAI tool call for '{name}' returned invalid JSON arguments."
                    ) from exc
                if not isinstance(arguments, dict):
                    raise OpenAIAdapterError(
                        f"OpenAI tool call for '{name}' returned non-object arguments."
                    )
                result = call_tool(name, arguments)
                output_items.append(
                    {
                        "type": "function_call_output",
                        "call_id": call_id,
                        "output": json.dumps(result),
                    }
                )

            response = self.create_response(
                model=model,
                input_items=output_items,
                tools=tools,
                previous_response_id=response.get("id"),
                max_output_tokens=max_output_tokens,
            )

        raise OpenAIAdapterError("OpenAI tool-calling loop exceeded max_round_trips.")

    def create_response(
        self,
        *,
        model: str,
        input_items: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
        previous_response_id: str | None = None,
        parallel_tool_calls: bool | None = None,
        max_output_tokens: int | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "model": model,
            "input": input_items,
        }
        if tools:
            payload["tools"] = tools
        if previous_response_id is not None:
            payload["previous_response_id"] = previous_response_id
        if parallel_tool_calls is not None:
            payload["parallel_tool_calls"] = parallel_tool_calls
        if max_output_tokens is not None:
            payload["max_output_tokens"] = max_output_tokens
        return self._post(payload)

    @staticmethod
    def _build_input(*, prompt: str, system: str | None) -> list[dict[str, Any]]:
        messages: list[dict[str, Any]] = []
        if system:
            messages.append(
                {
                    "role": "system",
                    "content": [{"type": "input_text", "text": system}],
                }
            )
        messages.append(
            {
                "role": "user",
                "content": [{"type": "input_text", "text": prompt}],
            }
        )
        return messages

    def _post(self, payload: dict[str, Any]) -> dict[str, Any]:
        body = json.dumps(payload).encode("utf-8")
        req = request.Request(
            url=f"{self.base_url.rstrip('/')}/responses",
            data=body,
            method="POST",
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
        )

        try:
            with request.urlopen(req, timeout=self.timeout_s) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise OpenAIAdapterError(f"OpenAI HTTP {exc.code}: {detail}") from exc
        except error.URLError as exc:
            raise OpenAIAdapterError(f"OpenAI network error: {exc.reason}") from exc
        except TimeoutError as exc:
            raise OpenAIAdapterError("OpenAI request timed out.") from exc
        except json.JSONDecodeError as exc:
            raise OpenAIAdapterError("OpenAI response was not valid JSON.") from exc


def _extract_text(payload: dict[str, Any]) -> str:
    direct = payload.get("output_text")
    if isinstance(direct, str) and direct.strip():
        return direct.strip()

    output = payload.get("output")
    if not isinstance(output, list):
        return ""

    fragments: list[str] = []
    for item in output:
        if not isinstance(item, dict):
            continue
        content = item.get("content")
        if not isinstance(content, list):
            continue
        for block in content:
            if not isinstance(block, dict):
                continue
            text = block.get("text")
            if isinstance(text, str) and text.strip():
                fragments.append(text.strip())

    return "\n".join(fragments).strip()


def _extract_function_calls(payload: dict[str, Any]) -> list[dict[str, str]]:
    output = payload.get("output")
    if not isinstance(output, list):
        return []

    function_calls: list[dict[str, str]] = []
    for item in output:
        if not isinstance(item, dict):
            continue
        if item.get("type") != "function_call":
            continue
        name = item.get("name")
        call_id = item.get("call_id")
        arguments = item.get("arguments")
        if all(isinstance(value, str) for value in [name, call_id, arguments]):
            function_calls.append(
                {
                    "name": name,
                    "call_id": call_id,
                    "arguments": arguments,
                }
            )
    return function_calls

File: src/adapters/openai.rs
use super::{
    AdapterError, CompletionRequest, JsonTransport, ModelClient, ToolExecutor, default_transport,
};
use serde_json::{Value, json};
use std::collections::BTreeMap;
use std::sync::Arc;
use std::time::Duration;

const PROVIDER: &str = "OpenAI";

pub struct OpenAiClient {
    api_key: String,
    base_url: String,
    transport: Arc<dyn JsonTransport>,
}

impl OpenAiClient {
    pub fn new(api_key: impl Into<String>) -> Result<Self, AdapterError> {
        Self::with_base_url(api_key, "https://api.openai.com/v1")
    }

    pub fn with_base_url(
        api_key: impl Into<String>,
        base_url: impl Into<String>,
    ) -> Result<Self, AdapterError> {
        Ok(Self {
            api_key: api_key.into(),
            base_url: base_url.into(),
            transport: default_transport(Duration::from_secs(45))?,
        })
    }

    #[cfg(test)]
    pub(crate) fn with_transport(transport: Arc<dyn JsonTransport>) -> Self {
        Self {
            api_key: "test".into(),
            base_url: "https://example.test/v1".into(),
            transport,
        }
    }

    fn response(&self, payload: Value) -> Result<Value, AdapterError> {
        self.transport.post(
            PROVIDER,
            &format!("{}/responses", self.base_url.trim_end_matches('/')),
            &BTreeMap::from([
                ("Authorization".into(), format!("Bearer {}", self.api_key)),
                ("Content-Type".into(), "application/json".into()),
            ]),
            &payload,
        )
    }
}

impl ModelClient for OpenAiClient {
    fn complete(&self, request: CompletionRequest<'_>) -> Result<String, AdapterError> {
        let mut payload =
            json!({"model":request.model,"input":input(request.prompt,request.system)});
        if let Some(max) = request.max_output_tokens {
            payload["max_output_tokens"] = max.into();
        }
        extract_text(&self.response(payload)?).ok_or(AdapterError::MissingText(PROVIDER))
    }

    fn complete_with_tools(
        &self,
        request: CompletionRequest<'_>,
        tools: &[Value],
        call_tool: &ToolExecutor,
        max_round_trips: usize,
    ) -> Result<String, AdapterError> {
        let mut payload = json!({"model":request.model,"input":input(request.prompt,request.system),"tools":tools,"parallel_tool_calls":false});
        if let Some(max) = request.max_output_tokens {
            payload["max_output_tokens"] = max.into();
        }
        let mut response = self.response(payload)?;
        for _ in 0..max_round_trips {
            let calls = function_calls(&response)?;
            if calls.is_empty() {
                return extract_text(&response).ok_or(AdapterError::MissingText(PROVIDER));
            }
            let mut outputs = Vec::new();
            for call in calls {
                let args: Value = serde_json::from_str(&call.arguments).map_err(|_| {
                    AdapterError::InvalidToolCall {
                        provider: PROVIDER,
                        tool: call.name.clone(),
                        problem: "invalid JSON arguments".into(),
                    }
                })?;
                let object = args
                    .as_object()
                    .ok_or_else(|| AdapterError::InvalidToolCall {
                        provider: PROVIDER,
                        tool: call.name.clone(),
                        problem: "non-object arguments".into(),
                    })?;
                let result = call_tool(&call.name, object)?;
                outputs.push(json!({"type":"function_call_output","call_id":call.call_id,"output":serde_json::to_string(&result).unwrap()}));
            }
            let mut next = json!({"model":request.model,"input":outputs,"tools":tools,"previous_response_id":response.get("id").and_then(Value::as_str)});
            if let Some(max) = request.max_output_tokens {
                next["max_output_tokens"] = max.into();
            }
            response = self.response(next)?;
        }
        Err(AdapterError::ToolLoopLimit(PROVIDER))
    }
}

fn input(prompt: &str, system: Option<&str>) -> Vec<Value> {
    let mut items = Vec::new();
    if let Some(system) = system {
        items.push(json!({"role":"system","content":[{"type":"input_text","text":system}]}))
    }
    items.push(json!({"role":"user","content":[{"type":"input_text","text":prompt}]}));
    items
}
fn extract_text(v: &Value) -> Option<String> {
    if let Some(x) = v
        .get("output_text")
        .and_then(Value::as_str)
        .filter(|x| !x.trim().is_empty())
    {
        return Some(x.trim().into());
    }
    let parts: Vec<_> = v
        .get("output")?
        .as_array()?
        .iter()
        .filter_map(|x| x.get("content")?.as_array())
        .flatten()
        .filter_map(|x| x.get("text")?.as_str())
        .filter(|x| !x.trim().is_empty())
        .map(str::trim)
        .collect();
    (!parts.is_empty()).then(|| parts.join("\n"))
}
struct FunctionCall {
    name: String,
    call_id: String,
    arguments: String,
}
fn function_calls(v: &Value) -> Result<Vec<FunctionCall>, AdapterError> {
    let mut out = Vec::new();
    for item in v
        .get("output")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        if item.get("type").and_then(Value::as_str) != Some("function_call") {
            continue;
        }
        let get = |key| item.get(key).and_then(Value::as_str).map(str::to_owned);
        if let (Some(name), Some(call_id), Some(arguments)) =
            (get("name"), get("call_id"), get("arguments"))
        {
            out.push(FunctionCall {
                name,
                call_id,
                arguments,
            })
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adapters::tests::FakeTransport;
    #[test]
    fn extracts_direct_and_nested_text() {
        assert_eq!(extract_text(&json!({"output_text":" hi "})).unwrap(), "hi");
        assert_eq!(
            extract_text(&json!({"output":[{"content":[{"text":"one"},{"text":"two"}]}]})).unwrap(),
            "one\ntwo"
        )
    }
    #[test]
    fn executes_tool_loop() {
        let fake = FakeTransport::new(vec![
            json!({"id":"r1","output":[{"type":"function_call","name":"lookup","call_id":"c1","arguments":"{\"q\":\"rust\"}"}]}),
            json!({"id":"r2","output_text":"done"}),
        ]);
        let client = OpenAiClient::with_transport(fake.clone());
        let text = client
            .complete_with_tools(
                CompletionRequest {
                    model: "gpt-test",
                    prompt: "go",
                    system: None,
                    max_output_tokens: None,
                },
                &[json!({"type":"function","name":"lookup","parameters":{"type":"object"}})],
                &|name, args| Ok(json!({"name":name,"q":args["q"]})),
                3,
            )
            .unwrap();
        assert_eq!(text, "done");
        assert_eq!(fake.requests.lock().unwrap().len(), 2)
    }
}
